privacyidea/lib/tokens/tantoken.py

```diff
@@ -0,0 +1,190 @@
+from privacyidea.lib.policy import SCOPE
+from privacyidea.lib import _
+from privacyidea.lib.policydecorators import libpolicy
+from privacyidea.lib.crypto import geturandom, hash
+import binascii
+
+log = logging.getLogger(__name__)
+DEFAULT_COUNT = 100
+SALT_LENGTH = 4
+
+
+class TANACTION(object):
+    TANTOKEN_COUNT = "papertoken_count"
+
+
+class TanTokenClass(PaperTokenClass):
+    """
+    The TAN token allows to print out the next e.g. 100 OTP values.
+    This sheet of paper can be used to authenticate and strike out the used
+    OTP values. In contrast to the paper token, the OTP values of the TAN token
+    can be used randomly.
+    """
+
+    @log_with(log)
+    def __init__(self, db_token):
+        """
+        This creates a new TAN token object from a DB token object.
+
+        :param db_token: instance of the orm db object
+        :type db_token:  orm object
+        """
+        TokenClass.__init__(self, db_token)
+        self.set_type(u"tan")
+        self.hKeyRequired = False
+
+    @staticmethod
+    def get_class_type():
+        """
+        return the token type shortname
+
+        :return: 'paper'
+        :rtype: string
+        """
+        return "tan"
+
+    @staticmethod
+    def get_class_prefix():
+        """
+        Return the prefix, that is used as a prefix for the serial numbers.
+        :return: PITN
+        """
+        return "PITN"
+
+    @staticmethod
+    @log_with(log)
+    def get_class_info(key=None, ret='all'):
+        """
+        returns a subtree of the token definition
+
+        :param key: subsection identifier
+        :type key: string
+        :param ret: default return value, if nothing is found
+        :type ret: user defined
+        :return: subsection if key exists or user defined
+        :rtype: dict or scalar
+        """
+        res = {'type': 'tan',
+               'title': 'TAN Token',
+               'description': 'TAN: TANs printed on a sheet '
+                              'of paper.',
+               'init': {},
+               'config': {},
+               'user':  ['enroll'],
+               # This tokentype is enrollable in the UI for...
+               'ui_enroll': ["admin", "user"],
+               'policy': {
+                   SCOPE.ENROLL: {
+                       TANACTION.TANTOKEN_COUNT: {
+                           "type": "int",
+                           "desc": _("The number of OTP values, which are "
+                                     "printed on the paper.")
+                       }
+                   }
+               }
+               }
+
+        if key:
+            ret = res.get(key, {})
+        else:
+            if ret == 'all':
+                ret = res
+        return ret
+
+    def update(self, param, reset_failcount=True):
+        tan_dict = {}
+        if "tans" in param:
+            # init tokens with tans
+            tans = param.get("tans").split()
+            tan_dict = {k: v for k, v in enumerate(tans)}
```

Comment:
In case we provide a list of TANs, we do not call the parent class' ``update`` implementation. I guess that's intentional because we do not need a OTP key and such, but this way we also lose the functionality of ``TokenClass.update``, which also sets the token description, PIN and token kind. This could be a problem, I think?

Reply:
hm. You are right. I did not want to add a useless otpkey. But maybe we should...
(The problem with the PIN and the different length is still open)

Reply:
done. Added the update method and also created an otplen from the tan list.